pkg/encoding/dictionary.go

```diff
@@ -0,0 +1,250 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package encoding
+
+import (
+	"bytes"
+	"fmt"
+	"math/bits"
+)
+
+// Dictionary is used for dictionary encoding.
+type Dictionary struct {
+	m       map[string]uint32
+	values  [][]byte
+	indices []uint32
+}
+
+// NewDictionary creates a dictionary.
+func NewDictionary() *Dictionary {
+	return &Dictionary{
+		m:       make(map[string]uint32),
+		values:  make([][]byte, 0),
+		indices: make([]uint32, 0),
+	}
+}
+
+// Reset resets the dictionary.
+func (d *Dictionary) Reset() {
+	clear(d.m)
+	d.values = d.values[:0]
+	d.indices = d.indices[:0]
+}
+
+// Add adds a value to the dictionary.
+func (d *Dictionary) Add(value []byte) {
```

Comment:
```suggestion
func (d *Dictionary) Add(value []byte) bool {
```

If the size of unique values exceeds a threshold(128 or 256?), returning false indicates that the block cannot use dict encoding.